`feat/multi-file` Added a toggle between the old gallery and new image list.

DeclineThyself opened the pull request:
Suggestion for #4719

This PR now only adds a toggle for the default view.

~~Although this seems to function, please consider this PR untested pseudo-code.~~
~~I prefer the old gallery over the new multi-image view, Please consider keeping both available through a toggle.~~

~~In this PR, I added a toggle to `Chat/Message Handling`.~~

~~Both `image_swipes` and `images` can now be displayed in either a galley or list view.~~

### TODO:

~~- `.mes_container` needs distinct styles for `images` and `image-swipes`.~~

#### Checklist:

- [X] I have read the [Contribution guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

DeclineThyself commented:
/gemini summary

DeclineThyself commented:
@gemini-code-assist missed this in the highlights. This will allow any extension referencing `message.extra.image` to select the current image from the correct array, instead of the first image.
This will also allow .set to update the selected image.

```javascript
//Index to store the currently selected Item.
let selectedIndex = plainProperty + 'Index';
obj[selectedIndex] ??= 0;

// If the plain property is already a getter, do nothing.
const hasGetterProperty = isGetterObjectProperty(obj, plainProperty);
if (hasGetterProperty) {
    return;
}

// Define the plain property as a getter/setter that wraps around the array property.
Object.defineProperty(obj, plainProperty, {
    // Getting the plain property returns the first item in the array property, or undefined if the array is empty.
    get: function () {
        console.trace(`Attempting to GET an array-wrapped property '${plainProperty}'. Use the array property '${arrayProperty}' instead.`);
        let longerArray = this?.[arrayProperty];
        if (oldArrayProperty) {
            longerArray = ((this?.[oldArrayProperty].length ?? 0) > (this?.[arrayProperty].length ?? 0)) ? this?.[oldArrayProperty] : this?.[arrayProperty];
        }
        return Array.isArray(longerArray) && longerArray.length > obj[selectedIndex] ? longerArray[obj[selectedIndex]] : void 0;
    },
    // Setting the plain property is not supported, as it would be ambiguous.
    set: function (path) {
        console.trace(`Attempting to SET an array-wrapped property '${plainProperty}'. Use the array property '${arrayProperty}' instead.`);
        let longerArray = this?.[arrayProperty];
        if (oldArrayProperty) {
            longerArray = ((this?.[oldArrayProperty].length ?? 0) > (this?.[arrayProperty].length ?? 0)) ? this?.[oldArrayProperty] : this?.[arrayProperty];
        }

        let index = longerArray.indexOf(path) > 0 ? longerArray.indexOf(path) : 0;
        this[selectedIndex] = index;
    },
    // Exclude the property from JSON serialization and from being listed in for...in loops.
    enumerable: false,
    // Make the property non-configurable to prevent deletion or redefinition.
    configurable: false,
});
 ````

DeclineThyself commented:
I keep forgetting that any @ mention will trigger a response from the bot.
It would be nice to have a way to turn that off.

Cohee1207 commented:
What is "old gallery"?

Cohee1207 commented:
> This will allow any extension

I can allow `set` to set the first array element (or add to an array if empty). I don't understand what is being suggested here.

Cohee1207 commented:
> What is "old gallery"?

JFYI, the "image swipes" system still exists, it will be displayed if you have only one image in the array. So you can actually do both multi-images and swipes at the same time if you really want.

DeclineThyself commented:
> JFYI, the "image swipes" system still exists, it will be displayed if you have only one image in the array. So you can actually do both multi-images and swipes at the same time if you really want.

I had thought that the image swipes system was being deprecated.

What's the difference between `image_swipes` and `images`?
It appears the SD extension can generate to both.

DeclineThyself commented:
> What is "old gallery"?

Regardless of whether there are `images` or`image_swipes`, I'd prefer an option to display them in the image carousel instead of the new view, which displays all the images at once.

Cohee1207 commented:
> I had thought that the image swipes system was being deprecated.

No, it still exists. And will exist for the time being, because my goal was to explicitly preserve this system. I guess what you _really_ meant is an ability to append files to `image_swipes` instead of `images` when choosing a message embed option. Like the image generation extension does with the paintbrush icon:

https://github.com/SillyTavern/SillyTavern/blob/769b24e3697377fd22b60356296507a8313bc448/public/scripts/extensions/stable-diffusion/index.js#L4246-L4266

> What's the difference between image_swipes and images?

If using image inlining, only the selected image swipe will be sent out to the model. If using multiple images, all images will be attached. image swipes are only displayed when you have exactly ONE image in the images array and image_swipes contain possible 'swaps' for this image that you can choose between.

Cohee1207 commented:
These two systems have their own distinct use cases: image swipes is needed to easily regenerate images generated with the Image Generation extension, multi-file is needed to send multiple images to the model when using image inlining. 

Believe it or not, but I tried to combine multi-file and swipes into a single system, but the outcome kinda sucked (#3703). It makes it more ambiguous what is actually is being sent out to the model. Especially for users who use ST as an image generation frontend (yes, they exist).

DeclineThyself commented:
Thank you, I understand now.

>I guess what you really meant is an ability to append files to image_swipes instead of images when choosing a message embed option. Like the image generation extension does with the paintbrush icon:

I don't see a need to append to `image_swipes` if you're not generating an image.

> I tried to combine multi-file and swipes into a single system... (https://github.com/SillyTavern/SillyTavern/pull/3703).

Thank you, I missed this.

I'll correct this PR, I was very confused between these two systems.

> These two systems have their own distinct use cases: image swipes is needed to easily regenerate images generated with the Image Generation extension, multi-file is needed to send multiple images to the model when using image inlining.

I don't see a need for these systems to be mutually exclusive, I'll propose a UI change.

Cohee1207 commented:
Could you outline your plan before attempting an implementation? Also, to your point, multi-file branch works exactly like it used to when you only have one image or just using image generation as usual. You won't notice any difference until you try to attach another file. The only thing that I can't guarantee is compatibility with all possible third-party extensions (in house extensions I can update myself), however I made sure that read-only operations are backwards compatible.

DeclineThyself commented:
> Could you outline your plan before attempting an implementation?

I'd outline my plan if I had one.
My goal is to display both `image-swipes` and `images`, with an option to toggle the carousel/gallery view.
Currently they're mutually exclusive.

DeclineThyself commented:
This took a while, I hope parts of this PR will be useful.

Let's see if the bot's summary can save me from writing an essay.
/gemini summary

DeclineThyself commented:
@Cohee1207, How's this so far?
If this PR isn't likely to be of use, I'd rather not spend more time on it.

Cohee1207 commented:
Is there an ability to choose between image swipes and multiple images per message? Or per chat? I think an ability to add image_swipes by upload would be better, to be honest.

DeclineThyself commented:
> Is there an ability to choose between image swipes and multiple images per message? Or per chat?

I added a `Toggle Gallery` setting to `Chat/Message Handling`.

> I think an ability to add image_swipes by upload would be better, to be honest.

I don't understand. In this PR both `extra.images and `extra.image_swipes` can be displayed as a gallery view or list view.

I can try to add a per-message gallery/list toggle.

To persist the data, I suggest altering `.extra.images` `.extra.files` and `.extra.videos` to be an object with information.
This would allow for extensions to persist data, and standardized display code.

```javascript
extra.images = {
  display = 'gallery'
  selectedIndex = 1,
    files = {
      [ title='Image Title', path='/image1.png', metadata={...},
      title='Image Title2', path='/image2.png', metadata={...}, ...]
   }
}

extra.videos = {
  display = 'list'
  selectedIndex = 0,
    files = {
      [ title='Video Title', path='/video1.wav', metadata={...},
      title='Video Title2', path='/video2.wav', metadata={...}, ...]
   }
}
```

DeclineThyself commented:
> Is there an ability to choose between image swipes and multiple images per message? Or per chat?

I added a persistent per-message gallery toggle, overriding the global setting.
https://github.com/SillyTavern/SillyTavern/pull/4722/commits/1b72eb5a0b1575c24e3fbf2a0fcc183dbd50ce4d
I also removed the redundant `Expand and zoom` Magnifying glass, clicking the image calls the same function.

I would prefer to make each file type independently toggle-able, but that would be messy to implement without the change I suggested.
> To persist the data, I suggest altering .extra.images .extra.files and .extra.videos to be an object with information.

DeclineThyself commented:
/gemini review

DeclineThyself commented on public/script.js:
```diff
@@ -1953,11 +1957,21 @@ export function ensureMessageMediaIsArray(mes) {
             get: function () {
                 console.trace(`Attempting to GET an array-wrapped property '${plainProperty}'. Use the array property '${arrayProperty}' instead.`);
-                return Array.isArray(this[arrayProperty]) && this[arrayProperty].length > 0 ? this[arrayProperty][0] : void 0;
+                return Array.isArray(this[arrayProperty]) && this[arrayProperty].length > obj[selectedIndex] ? arrayProperty[obj[selectedIndex]] : void 0;
```
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4722/commits/23e1d7a2974d2ae10795e21b41adb378f34306c8

DeclineThyself commented on public/scripts/chats.js:
```diff
@@ -965,41 +1010,35 @@ async function deleteMessageImage(messageId, imageIndex, messageBlock) {
-                message.extra.images.splice(imageIndex, 1);
-                delete message.extra.image_swipes;
-            }
+    if (imageIndex === 0 && Array.isArray(imageArray) && imageArray.length > 0) {
```
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4722/commits/23e1d7a2974d2ae10795e21b41adb378f34306c8

DeclineThyself commented on public/script.js:
```diff
@@ -1976,75 +1990,139 @@ export function ensureMessageMediaIsArray(mes) {
+    //Remove falsy from imagesArray.
+    imagesArray.forEach((image, index) => {
+        if (!image) { imagesArray.splice(index, 1); }
+    });
```
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4722/commits/23e1d7a2974d2ae10795e21b41adb378f34306c8

DeclineThyself commented on public/scripts/chats.js:
```diff
@@ -965,41 +1010,35 @@ async function deleteMessageImage(messageId, imageIndex, messageBlock) {
     } else {
-        message.extra.images.splice(imageIndex, 1);
+        imageArray.splice(imageIndex, 1);
+        message.extra[indexName] = imageIndex - 1;
```
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4722/commits/23e1d7a2974d2ae10795e21b41adb378f34306c8

DeclineThyself commented on public/scripts/extensions/stable-diffusion/index.js:
```diff
@@ -4323,17 +4325,12 @@ async function sdMessageButton(e) {
-            if (message.extra.images.length === 1) {
-                message.extra.image_swipes.push(image);
-            }
+            message.extra.imagesIndex = message.extra.images.push(image);
```
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4722/commits/23e1d7a2974d2ae10795e21b41adb378f34306c8

DeclineThyself commented:
/gemini review

DeclineThyself commented on public/script.js:
```diff
@@ -1976,75 +1990,140 @@ export function ensureMessageMediaIsArray(mes) {
-                });
-            }
+    let chatHeight = {};
+    //This there any cleaner way of passing a reference?
```
@gemini-code-assist
Answer the question posed in the comment.

DeclineThyself replied:
Removed comment.

DeclineThyself commented on public/script.js:
```diff
@@ -1953,11 +1957,21 @@ export function ensureMessageMediaIsArray(mes) {
+                // }
+                // else {
+                //     obj[selectedIndex] = this?.[arrayProperty].push(path);
+                // }
```
@gemini-code-assist
Answer the question posed in the comment.

DeclineThyself replied:
https://github.com/SillyTavern/SillyTavern/pull/4722/commits/5e4bcc412799c14e3f272a55abfb4f0b45e7e5df

Cohee1207 commented on public/index.html:
```diff
@@ -7267,9 +7271,11 @@ <h3><span data-i18n="Alternate Greetings" class="mdhotkey_location">Alternate Gr
+            <div class="mes_container mes_img_container">
                 <div class="mes_img_controls">
-                    <div title="Expand and zoom" class="right_menu_button fa-lg fa-solid fa-magnifying-glass mes_img_enlarge" data-i18n="[title]Expand and zoom"></div>
+                    <!-- Redundant? -->
```
No, it's not. This button opens an image full screen and enlarges it. Clicking on the image just enlarges it.

DeclineThyself replied:
Thank you.
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4722/commits/32c1ccb39eb988aac25943a6b923b00c5396e210

Cohee1207 commented:
Ok, new idea, hear me out: scrap image_swipes, and replace it with a per-message toggle of how the media is displayed (`extra.media_display: 'gallery'|'list'`) with a global power user toggle (gallery by default), and that extensions can choose how to display images, i.e. web search and image generation will use gallery by default. images stays a plain array of string, image_swipes auto-migrated to images and pre-set to gallery if it existed before. I think that solves both the redundancy and covers your wants and my dontwants.

Cohee1207 commented on public/script.js:
```diff
@@ -1953,11 +1957,24 @@ export function ensureMessageMediaIsArray(mes) {
+            set: function (path) {
                 console.trace(`Attempting to SET an array-wrapped property '${plainProperty}'. Use the array property '${arrayProperty}' instead.`);
+
+                let index = this?.[arrayProperty]?.indexOf(path);
```
I have a setter implementation stashed, but I have a question that basically expands on the "as it would be ambiguous" comment above: what to do if the set value is falsy? Options are: do nothing (already do, lol), remove the first array entry, empty the entire array. Only the last option carries the same semantic as falsying the legacy property, but I feel it's not correct and the developers of third-party extensions should go ahead and choose for themselves how they're gonna act instead of letting them inadvertently nuke the entire list of attached media. Sometimes, the only winning move is not to play.

DeclineThyself replied:
I fully agree with doing nothing on a falsy set.
I would prefer for an extension to be partially broken than for users to loose data.

In my opinion, many problems with ST today have been caused by old code that can't easily be changed due to extensions.

While the extension community as a whole is extremely valuable, I believe that better overall code and scheduled deprecations serves them better then a messy static code-base.

DeclineThyself commented:
> Ok, new idea, hear me out: scrap image_swipes, and replace it with a per-message toggle of how the media is displayed (`extra.media_display: 'gallery'|'list'`) with a global power user toggle (gallery by default), and that extensions can choose how to display images, i.e. web search and image generation will use gallery by default. images stays a plain array of string, image_swipes auto-migrated to images and pre-set to gallery if it existed before. I think that solves both the redundancy and covers your wants and my dontwants.

That seems perfect, I was very confused initially because I had assumed `image_swipes` would be deprecated.

Is there a reason that an object per image isn't possible?
I believe it would allow for future additions to be cleaner.
https://github.com/SillyTavern/SillyTavern/pull/4722#issuecomment-3473901025
```javascript
extra.images = {
  display = 'gallery'
  selectedIndex = 1,
    files = {
      [ title='Image Title', path='/image1.png', metadata={...},
      title='Image Title2', path='/image2.png', metadata={...}, ...]
   }
}

extra.videos = {
  display = 'list'
  selectedIndex = 0,
    files = {
      [ title='Video Title', path='/video1.wav', metadata={...},
      title='Video Title2', path='/video2.wav', metadata={...}, ...]
   }
}
```

Cohee1207 commented:
> Is there a reason that an object per image isn't possible?

Everything is possible, I just like to keep it as simple to work with as possible. I would maybe prefer an array of objects instead, as we already do with the `files` property for `images` and `videos`. It DOES keep it simple by allowing a plain isArray or length check in logic and expansible with whatever properties you may need for extensions (i.e. a per-image title is great for image generation extension).

Cohee1207 commented:
Ok, my perfectly crafted plan has a fatal flaw: how to track the currently selected media file without turning the property into an object type.

Cohee1207 commented:
Except maybe an `images_index` and `videos_index` properties at the top level of `extra` object 🤔

DeclineThyself commented:
> > Is there a reason that an object per image isn't possible?
> 
> Everything is possible, I just like to keep it as simple to work with as possible. I would maybe prefer an array of objects instead, as we already do with the `files` property for `images` and `videos`. It DOES keep it simple by allowing a plain isArray or length check in logic and expansible with whatever properties you may need for extensions (i.e. a per-image title is great for image generation extension).

My primacy concern is with "group" (files/images/videos) metadata.
Let's say the user wants videos displayed an in a list, and images in a gallery.

Then `.extra` would need a separate `.extra.videoDisplay` and `extra.imageDisplay`.
Then if a new file type was added later, for example 3d models, then extra would also need `extra.3dDisplay`.

The same goes for `.extra.selectedImage`  etc. (or `images_index`/`videosIndex`)

That's why I suggest keeping the "metadata" in the same object as the data. It results in far cleaner code when passing data between functions.

And I can use the `.attr('data-alternates-name')` trick (seen in `getContainerInfo`) to map the DOM element to it's object within `message.extra`.

DeclineThyself commented:
For example, this is how multiple `.mes_containers` can be standardized:
https://github.com/SillyTavern/SillyTavern/pull/4722/files#diff-a4e2bb42942b7497efce08c9278c128574525ac0ea1ce0d7fa231cdcd949f8dfR2211-R2220
```javascript
/**
 * Returns information about the closest .mes_container.
 * @returns {object}
 */
function getContainerInfo(containerClass){
    const messageBlock = $(this).closest('.mes');
    const messageId = Number(messageBlock.attr('mesid'));
    const mediaBlock = $(this).closest('.mes_container');
    const alternatesArrayName = String(mediaBlock.attr('data-alternates-name'));
    const alternatesIndexName = String(mediaBlock.attr('data-index-name'));
    const index = Number(mediaBlock.attr('data-index'));
    return { messageBlock, messageId, mediaBlock, alternatesArrayName, alternatesIndexName, index };
}
```

https://github.com/SillyTavern/SillyTavern/pull/4722/files#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R2008
```javascript
/**
 * Syncs the displayed images with the given imagesArrayName.
 * If the given array is blank, the container will be removed.
 * @param {object} mes Message object
 * @param {JQuery<HTMLElement>} messageElement Message element
 * @param {string} [imagesArrayName]
 * @param {string} [indexName]
 * @param {function} [doAdjustScroll] Function to run on image load.
 * @param {boolean} [useGallery] Toggle the gallery.
 */
export function updateMessageImages(mes, messageElement, imagesArrayName, indexName, doAdjustScroll, useGallery = power_user.toggle_gallery) {

    messageElement.find('.mes_text').toggleClass('displayNone', !mes.extra.inline_image);
    messageElement.find(`.mes_container[data-alternates-name=${imagesArrayName}]`).remove();
```



If it was stored in an object, then `getContainerInfo` would only need to return the object's name, and `updateMessageImages` would only need the object instead of the index and array names.

Cohee1207 commented:
But then you're leaving files a simple array property. Files are also a type of media but they don't need a gallery view.

DeclineThyself commented:
> But then you're leaving files a simple array property.

You're right, I accidentally swapped `{}` and `[]`

This is what I meant:

```javascript
extra.images = {
  display = 'gallery'
  selectedIndex = 1,
    files = [
      { title='Image Title', path='/image1.png', metadata={...},
      title='Image Title2', path='/image2.png', metadata={...}} ...]
   }
}

extra.files = {
  display = 'list'
  selectedIndex = 0,
    files = [
      [{ title='File Name', path='/file.zip', metadata={...}},
      {title='File Name2', path='/file2.txt', metadata={...}} ...]
}
```

Cohee1207 commented:
Having per-media type toggle between gallery / list seems redundant to me. Just one would be enough IMO. And possibly even combine videos / images into a single carousel (since image generation extension can also technically generate videos! but will be attaching them separately from images?).

Cohee1207 commented:
Call me a hater but I lowkey want to go with this. It just vibes with me better.

<img width="330" height="480" alt="image" src="https://github.com/user-attachments/assets/32df6f52-8a17-4b8e-932e-02fa3e857c68" />

DeclineThyself commented:
> Having per-media type toggle between gallery / list seems redundant to me. Just one would be enough IMO. And possibly even combine videos / images into a single carousel (since image generation extension can also technically generate videos! but will be attaching them separately from images?).

That can be handled in the interface.
Each object could have a group. (`extra.images.group`, `extra.videos.group`)

Then groups could be modified with `messageElement.find('.mes_container[data-group=${groupName}]')`

DeclineThyself commented:
> Call me a hater but I lowkey want to go with this. It just vibes with me better.

Ok, Thanks for hearing me out.
I still believe attachments should be unified.

I'll update my code to match yours.

Cohee1207 commented:
No need, I can handle it. I'm already on it.

Cohee1207 commented:
I did it? https://github.com/SillyTavern/SillyTavern/pull/4719/commits/17abd85610dcafc847d78c107d45563b9a71983d

Cohee1207 commented:
I wasn't sure if I need to write some code to migrate intermittent  'images'/'videos' array models to the new media format. So if you've lost a lot of attachments after switching to the PR branch - I'm sorry.

Commit:
Added "Toggle Gallery" button.
Added `getContainerInfo`.

DeclineThyself commented:
> I did it? [17abd85](https://github.com/SillyTavern/SillyTavern/commit/17abd85610dcafc847d78c107d45563b9a71983d)

Your code is far cleaner than mine. Here's what I believe may be worth keeping from my branch.
https://github.com/SillyTavern/SillyTavern/pull/4722/commits/ada4d640bc124a9a38ea10a32b41a712bac85acd

I added a global "Toggle Gallery" button that changes the default, and added `getContainerInfo` to unify some repeated code.

>I wasn't sure if I need to write some code to migrate intermittent 'images'/'videos' array models to the new media format. So if you've lost a lot of attachments after switching to the PR branch - I'm sorry.

No apologies needed, any data loss would have been recoverable from backups.

DeclineThyself commented:
/gemini review

Cohee1207 committed:
Refactor

Cohee1207 committed:
Change checkbox toggle to select

Cohee1207 committed:
Ensure media_display is set correctly only if any image_swipes were migrated

Cohee1207 committed:
Rename function

Cohee1207 committed:
Support Date in parseTimestamp

Cohee1207 committed:
Add type to main chat array

Cohee1207 committed:
Add media display reload prompt

Cohee1207 reviewed (approved):
I brought the toggle back to the UI Theme block, by made the reload to be a prompt, if any messages in the current chat are to be affected by the change.

DeclineThyself commented:
> I brought the toggle back to the UI Theme block, by made the reload to be a prompt, if any messages in the current chat are to be affected by the change.

Thank you!

Are you sure about this line? https://github.com/SillyTavern/SillyTavern/pull/4722/commits/db630be5f8332f737b192f0dc68e286e871dff7c#diff-4a4ec86e4c05b413e9b128d0f702c797e4a2afd8be513301c019824560d0c8e6R1981-L1985
If someone prefers the new list style, they would need to toggle every message in every chat.
Would it be better leave `obj.media_display` unset so `power_user.media_display` can take priority?

https://github.com/SillyTavern/SillyTavern/pull/4722/commits/c6179e602ea6ba8de23faa8041ed79009e3d810b#diff-c2cc24bc9001b11b6add48a4cd8f893d5d6c6e4d1bd254158bd14ab997f552cdR4675
Should this be `Default Media Style:` instead of `Media Style:`? Users may assume incorrectly and miss the title text.

Cohee1207 commented:
> Are you sure about this line?  

Yes, absolutely. For people who extensively used image swipes before, the display style should stay consistent after this update. Also extensions can set their own preferred display style, regardless of the user selection (can be overridden manually). The style selection is primarily for manual uploads as it will use a theme setting.
